Spy async bug introduced in Python 3.8.1

I believe [bpo-38875](https://bugs.python.org/issue38857) has introduced a bug in `mocker.spy()`. Spied coroutines are no longer awaited as they should be.

This test highlights the issue:

```python
import pytest


class MyClass:
    async def method(self):
        return "hi"


@pytest.mark.asyncio
async def test_spy_async(mocker):
    my_obj = MyClass()
    mocker.spy(my_obj, 'method')

    returned = await my_obj.method()

    # Passes
    assert my_obj.method.called

    # AssertionError: assert <coroutine object MyClass.method at 0x10f2a5d40> == 'hi'
    assert returned == "hi"
```

It looks like `spy()` (specifically, `wrapper()` defined within `spy()`) returns a coroutine and expects `unittest`'s `patch()` do to the awaiting, rather than doing the awaiting itself. This worked until bpo-38875 changed this behaviour. Now `unittest` will no longer await retuned awaitables, rather it returns them.

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -56,10 +56,11 @@ jobs:
 
   deploy:
 
+    if: github.event_name == 'push' && startsWith(github.event.ref, 'refs/tags')
+
     runs-on: ubuntu-latest
 
     needs: [build, linting]
-
 
     steps:
     - uses: actions/checkout@v1
@@ -75,7 +76,6 @@ jobs:
       run: |
         python setup.py sdist bdist_wheel
     - name: Publish package to PyPI
-      if: github.event_name == 'push' && startsWith(github.event.ref, 'refs/tags')
       uses: pypa/gh-action-pypi-publish@master
       with:
         user: __token__

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -1,3 +1,14 @@
+3.0.0 (UNRELEASED)
+------------------
+
+* Python 2.7 and 3.4 are no longer supported. Users using ``pip 9`` or later will install
+  a compatible version automatically.
+
+* ``mocker.spy`` now also works with ``async def`` functions (`#179`_). Thanks `@frankie567`_ for the PR!
+
+.. _#179: https://github.com/pytest-dev/pytest-mock/issues/179
+.. _@frankie567: https://github.com/frankie567
+
 2.0.0 (2020-01-04)
 ------------------
 

--- a/README.rst
+++ b/README.rst
@@ -114,6 +114,8 @@ In addition, spy objects contain two extra attributes:
 
 ``mocker.spy`` also works for class and static methods.
 
+As of version 3.0.0, ``mocker.spy`` also works with ``async def`` functions.
+
 .. note::
 
     In versions earlier than ``2.0``, the attributes were called ``return_value`` and

--- a/setup.py
+++ b/setup.py
@@ -19,7 +19,7 @@ setup(
     description="Thin-wrapper around the mock package for easier use with pytest",
     long_description=open("README.rst", encoding="utf-8").read(),
     keywords="pytest mock",
-    extras_require={"dev": ["pre-commit", "tox"]},
+    extras_require={"dev": ["pre-commit", "tox", "pytest-asyncio"]},
     classifiers=[
         "Development Status :: 5 - Production/Stable",
         "Framework :: Pytest",

--- a/src/pytest_mock/plugin.py
+++ b/src/pytest_mock/plugin.py
@@ -1,3 +1,4 @@
+import asyncio
 import functools
 import inspect
 
@@ -96,7 +97,6 @@ class MockFixture:
                 if isinstance(value, (classmethod, staticmethod)):
                     autospec = False
 
-        @functools.wraps(method)
         def wrapper(*args, **kwargs):
             spy_obj.spy_return = None
             spy_obj.spy_exception = None
@@ -109,7 +109,24 @@ class MockFixture:
                 spy_obj.spy_return = r
             return r
 
-        spy_obj = self.patch.object(obj, name, side_effect=wrapper, autospec=autospec)
+        async def async_wrapper(*args, **kwargs):
+            spy_obj.spy_return = None
+            spy_obj.spy_exception = None
+            try:
+                r = await method(*args, **kwargs)
+            except Exception as e:
+                spy_obj.spy_exception = e
+                raise
+            else:
+                spy_obj.spy_return = r
+            return r
+
+        if asyncio.iscoroutinefunction(method):
+            wrapped = functools.update_wrapper(async_wrapper, method)
+        else:
+            wrapped = functools.update_wrapper(wrapper, method)
+
+        spy_obj = self.patch.object(obj, name, side_effect=wrapped, autospec=autospec)
         spy_obj.spy_return = None
         spy_obj.spy_exception = None
         return spy_obj

--- a/tests/test_pytest_mock.py
+++ b/tests/test_pytest_mock.py
@@ -405,6 +405,21 @@ def test_callable_like_spy(testdir, mocker):
     assert spy.spy_return == 20
 
 
+@pytest.mark.asyncio
+async def test_instance_async_method_spy(mocker):
+    class Foo:
+        async def bar(self, arg):
+            return arg * 2
+
+    foo = Foo()
+    spy = mocker.spy(foo, "bar")
+
+    result = await foo.bar(10)
+
+    spy.assert_called_once_with(10)
+    assert result == 20
+
+
 @contextmanager
 def assert_traceback():
     """

--- a/tox.ini
+++ b/tox.ini
@@ -5,6 +5,7 @@ envlist = py{35,36,37,38}, linting, norewrite
 passenv = USER USERNAME
 deps =
     coverage
+    pytest-asyncio
 commands =
     coverage run --append --source={envsitepackagesdir}/pytest_mock -m pytest tests
 

